File: go/sort/mergesort.go
package sort

func merge(left, right []int64) []int64 {

	result := make([]int64, 0, len(left)+len(right))
	i, j := 0, 0

	for i < len(left) && j < len(right) {
		if left[i] < right[j] {
			result = append(result, left[i])
			i++
		} else {
			result = append(result, right[j])
			j++
		}
	}

	for i < len(left) {
		result = append(result, left[i])
		i++
	}

	for j < len(right) {
		result = append(result, right[j])
		j++
	}

	return result

}

func MergeSort(input []int64) []int64 {

	if len(input) <= 1 {
		return input
	}

	mid := len(input) / 2

	leftSide := input[0:mid]
	rightSide := input[mid:]

	leftSide = MergeSort(leftSide)
	rightSide = MergeSort(rightSide)

	return merge(leftSide, rightSide)
}

File: python/sort/mergesort.py
def merge(left, right):
    i=0
    j=0
    k = 0
    
    result = []
    while (i < len(left)) and (j < len(right)):
        if left[i] < right[j]:
            result.insert(k, left[i])
            i = i + 1
        else:
            result.insert(k, right[j])
            j = j +1

        k = k + 1
    
    while i < len(left):
        result.insert(k, left[i])
        i = i + 1
        k = k +1

    while j < len(right):
        result.insert(k, right[j])
        j = j + 1
        k = k +1

    return result
    
def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    
    mid = len(arr)//2
    left = arr[:mid]
    right = arr[mid:]

    left = merge_sort(left)
    right = merge_sort(right)
    return merge(left, right)




    